nano/node/transport/tcp_listener.cpp

```diff
@@ -1,281 +1,381 @@
-	acceptor.set_option (boost::asio::ip::tcp::acceptor::reuse_address (true));
-	boost::system::error_code ec;
-	acceptor.bind (local, ec);
-	if (!ec)
+	try
 	{
-		acceptor.listen (boost::asio::socket_base::max_listen_connections, ec);
+		acceptor.open (local.protocol ());
+		acceptor.set_option (asio::ip::tcp::acceptor::reuse_address (true));
+		acceptor.bind (local);
+		acceptor.listen (asio::socket_base::max_listen_connections);
+
+		logger.info (nano::log::type::tcp_listener, "Listening for incoming connections on: {}", fmt::streamed (acceptor.local_endpoint ()));
 	}
-	if (ec)
+	catch (boost::system::system_error const & ex)
 	{
-		node.logger.critical (nano::log::type::tcp_listener, "Error while binding for incoming TCP: {} (port: {})", ec.message (), acceptor.local_endpoint ().port ());
-		throw std::runtime_error (ec.message ());
+		logger.critical (nano::log::type::tcp_listener, "Error while binding for incoming TCP: {} (port: {})", ex.what (), port);
+
+		throw std::runtime_error (ex.code ().message ());
 	}
 
-	on_connection (callback_a);
+	thread = std::thread ([this] {
+		nano::thread_role::set (nano::thread_role::name::tcp_listener);
+		try
+		{
+			logger.debug (nano::log::type::tcp_listener, "Starting acceptor thread");
+			run ();
+			logger.debug (nano::log::type::tcp_listener, "Stopped acceptor thread");
+		}
+		catch (std::exception const & ex)
+		{
+			logger.critical (nano::log::type::tcp_listener, "Error: {}", ex.what ());
+			release_assert (false); // Should be handled earlier
+		}
+		catch (...)
+		{
+			logger.critical (nano::log::type::tcp_listener, "Unknown error");
+			release_assert (false); // Should be handled earlier
+		}
+	});
+
+	cleanup_thread = std::thread ([this] {
+		nano::thread_role::set (nano::thread_role::name::tcp_listener);
+		run_cleanup ();
+	});
 }
 
 void nano::transport::tcp_listener::stop ()
 {
-	decltype (connections) connections_l;
+	logger.info (nano::log::type::tcp_listener, "Stopping listeninig for incoming connections and closing all sockets...");
 	{
 		nano::lock_guard<nano::mutex> lock{ mutex };
 		stopped = true;
-		connections_l.swap (connections);
+
+		boost::system::error_code ec;
+		acceptor.close (ec); // Best effort to close the acceptor, ignore errors
+		if (ec)
+		{
+			logger.error (nano::log::type::tcp_listener, "Error while closing acceptor: {}", ec.message ());
+		}
 	}
+	condition.notify_all ();
 
-	nano::lock_guard<nano::mutex> lock{ mutex };
-	boost::asio::dispatch (strand, [this_l = shared_from_this ()] () {
-		this_l->acceptor.close ();
+	if (thread.joinable ())
+	{
+		thread.join ();
+	}
+	if (cleanup_thread.joinable ())
+	{
+		cleanup_thread.join ();
+	}
+
+	decltype (connections) connections_l;
+	{
+		nano::lock_guard<nano::mutex> lock{ mutex };
+		connections_l.swap (connections);
+	}
 
-		for (auto & address_connection_pair : this_l->connections_per_address)
+	for (auto & connection : connections_l)
+	{
+		if (auto socket = connection.socket.lock ())
```

Comment:
Supporting this, since tcp_server holds the shared_ptr and tcp_listener the weak_ptr, the tcp_server should clean up.

Reply:
Is it OK to modify the behaviour as requested and still keep servers as weak_ptrs?

If tcp_server owns and is responsible for closing the socket, then it would imply that tcp_listener should be responsible for keeping references to tcp_servers. This is not the case right now, servers are stored by weak_ptrs. I can't say with full certainty how those servers are stored and used in other parts of the node.

Reply:
I forgot that tcp_listener holds weak pointers to tcp_server and sockets.
So in that case, tcp_listener tracks tcp_servers and closes them but doesn't guarantee their liveness.
It is a little confusing that way and not easy to understand where the objects are held and what guarantees their liveness.
As it stands it seems messy but in the interests of moving ahead and cleaning up after we have fixed more pressing problems then I am OK with just adding a TODO to clean up later and clearly assign a component to be the owner of the pointers.

Reply:
Long term, or even now if that is what we decide, it seems natural for tcp_listener to be the owner of the tcp_server objects and tcp_server objects to own the socket objects.

Reply:
As I said, I intent to keep existing behavior with this refactoring as much as possible and the existing code is:
```
std::unordered_map<nano::transport::tcp_server *, std::weak_ptr<nano::transport::tcp_server>> connections;
std::multimap<boost::asio::ip::address, std::weak_ptr<socket>> connections_per_address;
```
Whether it's intuitive or not it was apparently reviewed and approved, and having worked with network code recently I noticed that trying to change too much too quickly often leads to subtle bugs.

Reply:
OK, then it is good enough as it is for now.